### PR TITLE
Add findns: directive to config

### LINKED ISSUE #1419: Support find_packages_ns from setup.cfg
Now that #1312 has been merged, closing #97, we support the `find_packages_ns` function in `setup.py`. We should also enable this functionality in `setup.cfg`.

I suggest adding a `find_namespace:` directive. The current `find:` directive is implemented [here](https://github.com/pypa/setuptools/blob/cca86c7f1d4040834c3265ccecdd9e21b4036df5/setuptools/config.py#L512), so I think that would probably be the place to add support for  namespace packages.

Brought to my attention by @agronholm  in https://github.com/pypa/setuptools/issues/97#issuecomment-404195899

### PR DESCRIPTION
**Pull request by silkentrance**
## Summary of changes

- extend documentation on new `findns:` directive
- realize new `findns:` directive

Closes #1419

### Pull Request Checklist
- [x] Changes have tests
- [x] News fragment added in changelog.d. See [documentation](http://setuptools.readthedocs.io/en/latest/developer-guide.html#making-a-pull-request) for details

### CODE REVIEW

**Comment by silkentrance** [comment]
@pganssle Have a look and tell me if that is the right direction. I so, I will continue with implementing required tests.

**Comment by pganssle** [comment]
Is there a reason it can't be `find_namespace`? I'm even dubious about the `find_packages_ns` name, since it seems too obscure to me, I'd rather it be `find_namespace_packages` or something.

**Comment by pganssle** [comment]
The `find_packages_ns` stuff is indeed a completely different discussion so don't worry about that.

**Comment by pganssle** [comment]
This seems like it will work, but the implementation might be a bit cleaner if you do this:

```python
directive = value.split(':', 1)[0]
if directive not in {'find', 'find_namespace'}:
    ...
```

**Comment by silkentrance** [comment]
@pganssle good point, but this will be a different commit

**Comment by silkentrance** [comment]
@pganssle note that with the latest commit I refrained from using string.split() and instead used value.strip().

File: docs/setuptools.txt [hunk 1]
```diff
@@ -2389,8 +2389,8 @@ Metadata and options are set in the config sections of the same name.
 * In some cases, complex values can be provided in dedicated subsections for
   clarity.
 
-* Some keys allow ``file:``, ``attr:``, and ``find:`` directives in order to
-  cover common usecases.
+* Some keys allow ``file:``, ``attr:``, and ``find:`` and ``find_namespace:`` directives i
```
Comment (agronholm) [inline docs/setuptools.txt]: Typo at the end?

File: setuptools/config.py [hunk 2]
```diff
@@ -515,21 +515,29 @@ def _parse_packages(self, value):
...
         :rtype: list
         """
-        find_directive = 'find:'
+        find_directives = ['find:', 'find_namespace:']
+        trimmed_value = value.strip()
 
-        if not value.startswith(find_directive):
-            return self._parse_list(value)
+        if not trimmed_value in find_directives:
+          return self._parse_list(value)
```
Comment (agronholm) [inline setuptools/config.py]: Bad indentation

File: setuptools/config.py [hunk 3]
```diff
@@ -515,21 +515,29 @@ def _parse_packages(self, value):
...
+        if findns:
+          from setuptools import find_packages_ns as find_packages
+        else:
+          from setuptools import find_packages
 
         return find_packages(**find_kwargs)
 
     def parse_section_packages__find(self, section_options):
-        """Parses `packages.find` configuration file section.
+        """Parses `packages.find[ns]` configuration file section.
```
Comment (agronholm) [inline setuptools/config.py]: Does this really use a `packages.findns` section? The tests do not.

File: setuptools/tests/test_config.py [hunk 4]
```diff
@@ -596,6 +601,59 @@ def test_find_directive(self, tmpdir):
...
+        config.write(
+            '[options]\n'
+            'packages = find_namespace:\n'
+            '\n'
+            '[options.packages.find]\n'
+            'exclude =\n'
+            '    fake_package.sub_one\n'
+        )
+        with get_dist(tmpdir) as dist:
+            assert set(dist.packages) == set(
```
Comment (agronholm) [inline setuptools/tests/test_config.py]: You can use a set literal here. Python 2.6 support was dropped in v37.

File: setuptools/config.py (hunk 2 above)
Comment (silkentrance) [reply setuptools/config.py]: and there was even more where this came from...

Comment (silkentrance) [reply setuptools/config.py]: done

File: docs/setuptools.txt (hunk 1 above)
Comment (silkentrance) [reply docs/setuptools.txt]: done

File: setuptools/tests/test_config.py (hunk 4 above)
Comment (silkentrance) [reply setuptools/tests/test_config.py]: The other tests also use the non set literal, should I proceed?

Comment (silkentrance) [reply setuptools/tests/test_config.py]: done

**Comment by silkentrance** [comment]
If you comply to the changes, I will proceed with fixing #1421 based upon these changes.

File: setuptools/tests/test_config.py [hunk 5]
```diff
@@ -4,18 +4,23 @@
 from mock import patch
 from setuptools.dist import Distribution, _Distribution
 from setuptools.config import ConfigHandler, read_configuration
+from setuptools.extern.six import PY2, PY3
 
+py2_only = pytest.mark.xfail(not PY2, reason="Test runs on Python 2 only")
```
Comment (pganssle) [inline setuptools/tests/test_config.py]: I think these should be `pytest.mark.skip`, not `xfail`.

Comment (silkentrance) [reply setuptools/tests/test_config.py]: will have to fix the tests for find_packages_ns as well. you would want to fix the other tests that use that pattern.

Comment (agronholm) [reply setuptools/tests/test_config.py]: I concur.

Comment (silkentrance) [reply setuptools/tests/test_config.py]: six.PY2 seems to fail when marking tests for specific versions

```
SKIP [1] setuptools/tests/test_config.py:604: Test runs on Python 2 only
```

Comment (pganssle) [reply setuptools/tests/test_config.py]: I think that's indicating that the test was skipped. When you mark something `xfail`, it actually *runs* the test and ignores the result if it fails (or if you make it a strict xfail, actually fails the test suite if the test passes). `skip` doesn't run the test at all, so I believe that message is expected.

You are right that we should be switching many `xfail`s over to `skip`. I was planning on doing that at some point, but it takes time to figure out which ones are intended as an actual `xfail` (which generally means "this should work but it's a known bug - when this test starts passing drop this decorator") and which are just using the wrong decorator for `skip`.

**Comment by silkentrance** [comment]
currently working on py2_only/py3_only for test_config. It does not seem to work as expected, reporting false results.

**Comment by pganssle** [comment]
I pushed an updated version with the proper decorator to your branch.

**Comment by silkentrance** [comment]
@pganssle thanks. I did find out about skipif in the morning, too. Yesterday I was way too tired...

**Comment by silkentrance** [comment]
I have refactored out the markers for py2/py3 into a separate module since they had been redeclared. Other test modules can be refactored, too, but I leave it at that for the time being.

**Pull request closed without merging**